src/kernels/MIOpenMultilabelSoftMarginLoss.cpp

```diff
@@ -0,0 +1,130 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+__device__ FLOAT_ACCUM sigmoid(FLOAT_ACCUM x) { return 1 / (1 + exp(-x)); }
+__device__ FLOAT_ACCUM calc_loss(FLOAT_ACCUM x, FLOAT_ACCUM y)
+{
+    FLOAT_ACCUM sig = sigmoid(x);
+    return y * log(sig) + (1 - y) * log(1 - sig);
+}
+
+template <typename DTYPE>
+__device__ void multilabelsoftmarginlossunreducedforward2d(const DTYPE* __restrict__ I,
+                                                           const DTYPE* __restrict__ T,
+                                                           const DTYPE* __restrict__ W,
+                                                           DTYPE* __restrict__ O,
+                                                           tensor_view_t<2> I_tv,
+                                                           tensor_view_t<2> T_tv,
+                                                           tensor_view_t<1> W_tv,
+                                                           tensor_view_t<1> O_tv)
+{
+    const uint64_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+
+    size_t N = I_tv.size[0], C = I_tv.size[1];
+    size_t n = gid;
+    if(n >= N)
+        return;
+
+    FLOAT_ACCUM loss = 0;
+
+    // TODO: maybe faster reduce sum here
+    for(size_t c = 0; c < C; c++)
+    {
+        FLOAT_ACCUM w = CVT_FLOAT2ACCUM(W[W_tv.get_tensor_view_idx({c})]);
+        FLOAT_ACCUM i = CVT_FLOAT2ACCUM(I[I_tv.get_tensor_view_idx({n, c})]);
+        FLOAT_ACCUM t = CVT_FLOAT2ACCUM(T[T_tv.get_tensor_view_idx({n, c})]);
+
+        loss += -w * calc_loss(i, t);
+    }
+
+    O[O_tv.get_tensor_view_idx({n})] = CVT_ACCUM2FLOAT(loss / C);
+}
+
+extern "C" __global__ void
+MultilabelSoftMarginLossUnreducedForward2d(const INPUT_TYPE* __restrict__ I,
```

Comment:
If you are not going to have input and output types that differs, imo you can simply use `FLOAT` as the type for arguments. It should then be defined automatically by the compiler.

Reply:
@seungmanhan How do you think about this, should we only use FLOAT for kernel that have same type for input type and output type? I saw in the your last PR in ROCm/MIOpen https://github.com/ROCm/MIOpen/pull/2833 , in AddLayerNormForward you verify type of xDesc and yDesc need to be the same, but you still use INPUT_TYPE and OUTPUT_TYPE in AddLayernormFwdContiguous. Are there any specific reasons to do that?

Reply:
You don't have to worry about the previous work because it was done by referring to other miopen codes. In fact, the rules are something we continue to make. In fact, since most kernels in moDNN use the same input and output types, I think it is better to work according to o2buzzle's opinion.

Reply:
Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/e9f959f5c1bfb65a1fb4033449671487b45c9bf0)